### PR TITLE
[CINFRA-218] Construct Leader with Participant Flags

### PR DESCRIPTION
**Pull request by apetenchea**
### Scope & Purpose

The leader used to ignore participant flags. The generation is now taken into account, which led to a lot of tests being modified.
A test for checking that leadership cannot be established with an excluded follower has been added in `EstablishLeadershipTest.cpp`.
A potential bug was fixed in `TestReplicatedLog::becomeLeader`, increasing `replicationFactor` by 1, to take the leader into account.

- [x] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [x] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**

### CODE REVIEW

File: tests/Replication2/ReplicatedLog/LogStatusTest.cpp [hunk 1]
```diff
@@ -41,10 +41,18 @@ TEST_F(ReplicatedLogTest, quick_status_compare) {
...
+          .participants = {{leaderId, {}}, {followerId, {}}},
+      });
+
   auto leader = LogLeader::construct(
-      LoggerContext(Logger::REPLICATION2), _logMetricsMock, _optionsMock,
-      leaderId, std::move(coreA), LogTerm{1},
-      std::vector<std::shared_ptr<AbstractFollower>>{follower}, 2);
+      config, std::move(coreA), {follower}, std::move(participantsConfig),
+      leaderId, LogTerm{1}, LoggerContext(Logger::REPLICATION2),
+      _logMetricsMock, _optionsMock);
```
[the remark this answers is not in the document]
Comment (apetenchea) [reply tests/Replication2/ReplicatedLog/LogStatusTest.cpp]: Apart from `ReplicatedLogTest`, a similar function has been added to `LogMultiplexerTestBase`.

File: tests/Replication2/ReplicatedLog/SimpleInsertTests.cpp [hunk 2]
```diff
@@ -45,10 +45,8 @@ TEST_F(ReplicatedLogTest, write_single_entry_to_follower) {
   auto follower = std::make_shared<DelayedFollowerLog>(
       defaultLogger(), _logMetricsMock, followerId, std::move(coreB),
       LogTerm{1}, leaderId);
-  auto leader = LogLeader::construct(
-      LoggerContext(Logger::REPLICATION2), _logMetricsMock, _optionsMock,
-      leaderId, std::move(coreA), LogTerm{1},
-      std::vector<std::shared_ptr<AbstractFollower>>{follower}, 2);
+  auto leader =
+      createLeader(leaderId, LogTerm{1}, std::move(coreA), {follower}, 2);
```
Comment (maierlars) [inline tests/Replication2/ReplicatedLog/SimpleInsertTests.cpp:49]: Sorry for being picky but maybe we want to call the function `createLeaderWithDefaultFlags`?

File: tests/Replication2/Streams/TestLogSpecification.h [hunk 3]
```diff
@@ -59,6 +59,27 @@ struct LogMultiplexerTestBase
                                    test::MockLog>(id);
   }
 
+  static auto createLeader(
```
Comment (maierlars) [inline tests/Replication2/Streams/TestLogSpecification.h:62]: Same here?

**Review by maierlars — APPROVED** [review]

File: tests/Replication2/ReplicatedLog/SimpleInsertTests.cpp (hunk 2 above)
Comment (apetenchea) [reply tests/Replication2/ReplicatedLog/SimpleInsertTests.cpp:49]: Yes, that's much better name.

**Pull request merged**